Tests/SourceKitTests/CrashRecoveryTests.swift

```diff
@@ -0,0 +1,269 @@
+    self.wait(for: [semanticFunctionalityRestored], timeout: 30)
+
+    // Check that we get the same hover response from the in-memory state
+
+    XCTAssertNoThrow(try {
+      let postCrashHoverResponse = try ws.sk.sendSync(hoverRequest)
+      XCTAssertEqual(postCrashHoverResponse, expectedHoverResponse)
+    }())
+  }
+  
+  /// Crashes clangd and waits for it to restart
+  /// - Parameters:
+  ///   - ws: The workspace for which the clangd server shall be crashed
+  ///   - loc: A test location that points to the first line of a given file. This line needs to be otherwise empty.
+  private func crashClangd(for ws: SKTibsTestWorkspace, firstLineLoc loc: TestLocation) {
+    let clangdServer = ws.testServer.server!.languageService(for: loc.docUri, .cpp, in: ws.testServer.server!.workspace!)!
+    
+    let clangdCrashed = self.expectation(description: "clangd crashed")
+    let clangdRestarted = self.expectation(description: "clangd restarted")
+
+    clangdServer.addStateChangeHandler { (oldState, newState) in
+      switch newState {
+      case .connectionInterrupted:
+        clangdCrashed.fulfill()
+      case .connected:
+        clangdRestarted.fulfill()
+      default:
+        break
+      }
+    }
+
+    // Add a pragma to crash clang
+    let addCrashPragma = TextDocumentContentChangeEvent(range: loc.position..<loc.position, rangeLength: 0, text: "#pragma clang __debug crash\n")
+    ws.sk.send(DidChangeTextDocumentNotification(textDocument: VersionedTextDocumentIdentifier(loc.docUri, version: 3), contentChanges: [addCrashPragma]))
+
+    self.wait(for: [clangdCrashed], timeout: 5)
+
+    // Once clangds has crashed, remove the pragma again to allow it to restart
+    let removeCrashPragma = TextDocumentContentChangeEvent(range: loc.position..<Position(line: 1, utf16index: 0), rangeLength: 28, text: "")
+    ws.sk.send(DidChangeTextDocumentNotification(textDocument: VersionedTextDocumentIdentifier(loc.docUri, version: 4), contentChanges: [removeCrashPragma]))
+
+    self.wait(for: [clangdRestarted], timeout: 30)
+  }
+
+  func testClangdCrashRecovery() {
+    guard longTestsEnabled else {
+      return
+    }
+
+    let ws = try! staticSourceKitTibsWorkspace(name: "ClangCrashRecovery")!
+    let loc = ws.testLoc("loc")
+
+    try! ws.openDocument(loc.url, language: .cpp)
+
+    // Make a change to the file that's not saved to disk. This way we can check that we re-open the correct in-memory state.
+
+    let addFuncChange = TextDocumentContentChangeEvent(range: loc.position..<loc.position, rangeLength: 0, text: """
+
+      void main() {
+      }
+      """)
+    ws.sk.send(DidChangeTextDocumentNotification(textDocument: VersionedTextDocumentIdentifier(loc.docUri, version: 2), contentChanges: [addFuncChange]))
+
+    // Do a sanity check and verify that we get the expected result from a hover response before crashing clangd.
+
+    let expectedHoverContent = """
+      Declared in global namespace
```

Comment:
This text seems to vary between recent clangd versions - I tried this test locally and it failed because "loc" wasn't included at the bottom.  We generally try to keep our tests passing with both the latest toolchain and the most recent _released_ toolchain.  Let's change this only compare the range, or if you really want to check the markup content, just check a prefix like "Declared in".